RecoLocalMuon/GEMSegment/plugins/ME0SegAlgoMM.cc

```diff
@@ -0,0 +1,320 @@
+    return segments;
+  }
+}
+
+
+// ********************************************************************;
+ME0SegAlgoMM::ProtoSegments 
+ME0SegAlgoMM::clusterHits(const EnsembleHitContainer & rechits) {
+
+  ProtoSegments rechits_clusters; // this is a collection of groups of rechits
+
+  float dXclus_box = 0.0;
+  float dYclus_box = 0.0;
+
+  EnsembleHitContainer temp;
+  ProtoSegments seeds;
+
+  std::vector<float> running_meanX;
+  std::vector<float> running_meanY;
+
+  std::vector<float> seed_minX;
+  std::vector<float> seed_maxX;
+  std::vector<float> seed_minY;
+  std::vector<float> seed_maxY;
+
+  // split rechits into subvectors and return vector of vectors:
+  // Loop over rechits 
+  // Create one seed per hit
+  for(unsigned int i = 0; i < rechits.size(); ++i) {
+    temp.clear();
+    temp.push_back(rechits[i]);
+    seeds.push_back(temp);
+
+    // First added hit in seed defines the mean to which the next hit is compared
+    // for this seed.
+
+    running_meanX.push_back( rechits[i]->localPosition().x() );
+    running_meanY.push_back( rechits[i]->localPosition().y() );
+	
+    // set min/max X and Y for box containing the hits in the precluster:
+    seed_minX.push_back( rechits[i]->localPosition().x() );
+    seed_maxX.push_back( rechits[i]->localPosition().x() );
+    seed_minY.push_back( rechits[i]->localPosition().y() );
+    seed_maxY.push_back( rechits[i]->localPosition().y() );
+  }
+    
+  // merge clusters that are too close
+  // measure distance between final "running mean"
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    for(size_t MMM = NNN+1; MMM < seeds.size(); ++MMM) {
+      if(running_meanX[MMM] == running_max || running_meanX[NNN] == running_max ) {
+	LogDebug("ME0SegAlgoMM") << "[ME0SegAlgoMM::clusterHits]: ALARM! Skipping used seeds, this should not happen - inform developers!";
+	continue; //skip seeds that have been used 
+      }
+	  
+      // calculate cut criteria for simple running mean distance cut:
+      //dXclus = fabs(running_meanX[NNN] - running_meanX[MMM]);
+      //dYclus = fabs(running_meanY[NNN] - running_meanY[MMM]);
+      // calculate minmal distance between precluster boxes containing the hits:
+      if ( running_meanX[NNN] > running_meanX[MMM] ) dXclus_box = seed_minX[NNN] - seed_maxX[MMM];
+      else                                           dXclus_box = seed_minX[MMM] - seed_maxX[NNN];
+      if ( running_meanY[NNN] > running_meanY[MMM] ) dYclus_box = seed_minY[NNN] - seed_maxY[MMM];
+      else                                           dYclus_box = seed_minY[MMM] - seed_maxY[NNN];
+	  
+	  
+      if( dXclus_box < dXclusBoxMax && dYclus_box < dYclusBoxMax ) {
+	// merge clusters!
+	// merge by adding seed NNN to seed MMM and erasing seed NNN
+	    
+	// calculate running mean for the merged seed:
+	running_meanX[MMM] = (running_meanX[NNN]*seeds[NNN].size() + running_meanX[MMM]*seeds[MMM].size()) / (seeds[NNN].size()+seeds[MMM].size());
```

Comment:
seeds[MMM].size()  and that for NNN are never zero  together, right?
if not, add a protection.